File: src/frontmatter.ts
import type { Metadata } from "./model.js";

export interface FrontmatterSplit {
  metadata: Metadata;
  body: string;
}

export function splitFrontmatter(text: string): FrontmatterSplit {
  const lines = text.split(/\r?\n/);
  if (lines[0]?.trim() !== "---") {
    return { metadata: {}, body: text };
  }

  const end = lines.findIndex((line, index) => index > 0 && line.trim() === "---");
  if (end < 0) {
    return { metadata: {}, body: text };
  }

  const metadata: Metadata = {};
  for (const rawLine of lines.slice(1, end)) {
    const match = /^([A-Za-z0-9_-]+)\s*:\s*(.*)$/.exec(rawLine);
    if (!match) {
      continue;
    }
    metadata[match[1]] = stripQuotes(match[2].trim());
  }

  return { metadata, body: lines.slice(end + 1).join("\n") };
}

function stripQuotes(value: string): string {
  if (
    (value.startsWith('"') && value.endsWith('"')) ||
    (value.startsWith("'") && value.endsWith("'"))
  ) {
    return value.slice(1, -1);
  }
  return value;
}

File: markdown_gate/frontmatter.py
from __future__ import annotations

from typing import Any


def split_frontmatter(text: str) -> tuple[dict[str, Any], str]:
    """Parse a small YAML-like frontmatter block without adding dependencies."""
    if not text.startswith("---\n"):
        return {}, text

    lines = text.splitlines(keepends=True)
    end_index = None
    for idx in range(1, len(lines)):
        if lines[idx].strip() == "---":
            end_index = idx
            break

    if end_index is None:
        return {}, text

    metadata: dict[str, Any] = {}
    for raw in lines[1:end_index]:
        stripped = raw.strip()
        if not stripped or stripped.startswith("#") or ":" not in stripped:
            continue
        key, value = stripped.split(":", 1)
        metadata[key.strip()] = _parse_scalar(value.strip())

    # Preserve source line numbers after stripping metadata from semantic scans.
    body = "\n" * (end_index + 1) + "".join(lines[end_index + 1 :])
    return metadata, body


def _parse_scalar(value: str) -> Any:
    if value in {"true", "True"}:
        return True
    if value in {"false", "False"}:
        return False
    if len(value) >= 2 and value[0] == value[-1] and value[0] in {'"', "'"}:
        return value[1:-1]
    if value.startswith("[") and value.endswith("]"):
        inner = value[1:-1].strip()
        if not inner:
            return []
        return [_parse_scalar(part.strip()) for part in inner.split(",")]
    return value
